backupccl: TestProtectedTimestampsDuringBackup is flaky on master

On 82294d100ad1cb26b488f53e7241242d474d58d7, I ran
```
roachprod create lucy-stress
make roachprod-stress CLUSTER=lucy-stress STRESSFLAGS='-p 16' PKG=./pkg/ccl/backupccl TESTS=TestProtectedTimestampsDuringBackup
```
 and got this failure within a few hundred runs:
```
--- FAIL: TestProtectedTimestampsDuringBackup (57.99s)
    backup_test.go:3673: condition failed to evaluate within 45s: expected (?s)shouldQueue=true.*processing replica.*GC score after GC in trace:      0.000ms      0.000ms    === operation:manual gc queue run node:1 range:33/1:/{Table/52-Max} store:1 unfinished:
             0.012ms      0.012ms    event:[n1,s1,r33/1:/{Table/52-Max}] running gc.shouldQueue
             0.038ms      0.026ms    event:[n1,s1,r33/1:/{Table/52-Max}] shouldQueue=false, priority=0.000318


        goroutine 45 [running]:
        runtime/debug.Stack(0xc0041d3c60, 0x481bc60, 0xc002b13930)
                /usr/local/go/src/runtime/debug/stack.go:24 +0x9d
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x4913400, 0xc0005c5100, 0xc0041d3c60)
                /go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:37 +0x6b
        github.com/cockroachdb/cockroach/pkg/ccl/backupccl_test.TestProtectedTimestampsDuringBackup(0xc0005c5100)
                /go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/backup_test.go:3673 +0x802
        testing.tRunner(0xc0005c5100, 0x41d5550)
                /usr/local/go/src/testing/testing.go:909 +0xc9
        created by testing.(*T).Run
                /usr/local/go/src/testing/testing.go:960 +0x350
```
Full logs: [out.1586960612.gz](https://github.com/cockroachdb/cockroach/files/4481811/out.1586960612.gz)

--- a/pkg/ccl/backupccl/backup_job.go
+++ b/pkg/ccl/backupccl/backup_job.go
@@ -541,6 +541,14 @@ func (b *backupResumer) Resume(
 	}
 	b.deleteCheckpoint(ctx, p.ExecCfg())
 
+	if ptsID != nil && !b.testingKnobs.ignoreProtectedTimestamps {
+		if err := p.ExecCfg().DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+			return b.releaseProtectedTimestamp(ctx, txn, p.ExecCfg().ProtectedTimestampProvider)
+		}); err != nil {
+			log.Errorf(ctx, "failed to release protected timestamp: %v", err)
+		}
+	}
+
 	resultsCh <- tree.Datums{
 		tree.NewDInt(tree.DInt(*b.job.ID())),
 		tree.NewDString(string(jobs.StatusSucceeded)),
@@ -548,14 +556,6 @@ func (b *backupResumer) Resume(
 		tree.NewDInt(tree.DInt(res.Rows)),
 		tree.NewDInt(tree.DInt(res.IndexEntries)),
 		tree.NewDInt(tree.DInt(res.DataSize)),
-	}
-
-	if ptsID != nil && !b.testingKnobs.ignoreProtectedTimestamps {
-		if err := p.ExecCfg().DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
-			return b.releaseProtectedTimestamp(ctx, txn, p.ExecCfg().ProtectedTimestampProvider)
-		}); err != nil {
-			log.Errorf(ctx, "failed to release protected timestamp: %v", err)
-		}
 	}
 
 	// Collect telemetry.

--- a/pkg/ccl/backupccl/backup_test.go
+++ b/pkg/ccl/backupccl/backup_test.go
@@ -3626,7 +3626,14 @@ func TestProtectedTimestampsDuringBackup(t *testing.T) {
 	rowCount := runner.QueryStr(t, "SELECT * FROM foo")
 
 	go func() {
-		runner.Exec(t, `BACKUP TABLE FOO TO 'nodelocal://1/foo'`)
+		// N.B. We use the conn rather than the runner here since the test may
+		// finish before the job finishes. The test will finish as soon as the
+		// timestamp is no longer protected. If the test starts tearing down the
+		// cluster before the backup job is done, the test may still fail when the
+		// backup fails. This test does not particularly care if the BACKUP
+		// completes with a success or failure, as long as the timestamp is released
+		// shortly after the BACKUP is unblocked.
+		_, _ = conn.Exec(`BACKUP TABLE FOO TO 'nodelocal://1/foo'`) // ignore error.
 	}()
 
 	var jobID string
@@ -3669,11 +3676,15 @@ func TestProtectedTimestampsDuringBackup(t *testing.T) {
 
 	// Wait for the ranges to learn about the removed record and ensure that we
 	// can GC from the range soon.
-	gcRanRE := regexp.MustCompile("(?s)shouldQueue=true.*processing replica.*GC score after GC")
+	// This regex matches when all float priorities other than 0.00000. It does
+	// this by matching either a float >= 1 (e.g. 1230.012) or a float < 1 (e.g.
+	// 0.000123).
+	matchNonZero := "[1-9]\\d*\\.\\d+|0\\.\\d*[1-9]\\d*"
+	nonZeroProgressRE := regexp.MustCompile(fmt.Sprintf("priority=(%s)", matchNonZero))
 	testutils.SucceedsSoon(t, func() error {
 		writeGarbage(3, 10)
-		if trace := gcTable(false /* skipShouldQueue */); !gcRanRE.MatchString(trace) {
-			return fmt.Errorf("expected %v in trace: %v", gcRanRE, trace)
+		if trace := gcTable(false /* skipShouldQueue */); !nonZeroProgressRE.MatchString(trace) {
+			return fmt.Errorf("expected %v in trace: %v", nonZeroProgressRE, trace)
 		}
 		return nil
 	})

--- a/pkg/ccl/importccl/import_into_test.go
+++ b/pkg/ccl/importccl/import_into_test.go
@@ -147,11 +147,15 @@ func TestProtectedTimestampsDuringImportInto(t *testing.T) {
 
 	// Wait for the ranges to learn about the removed record and ensure that we
 	// can GC from the range soon.
-	gcRanRE := regexp.MustCompile("(?s)shouldQueue=true.*processing replica.*GC score after GC")
+	// This regex matches when all float priorities other than 0.00000. It does
+	// this by matching either a float >= 1 (e.g. 1230.012) or a float < 1 (e.g.
+	// 0.000123).
+	matchNonZero := "[1-9]\\d*\\.\\d+|0\\.\\d*[1-9]\\d*"
+	nonZeroProgressRE := regexp.MustCompile(fmt.Sprintf("priority=(%s)", matchNonZero))
 	testutils.SucceedsSoon(t, func() error {
 		writeGarbage(3, 10)
-		if trace := gcTable(false /* skipShouldQueue */); !gcRanRE.MatchString(trace) {
-			return fmt.Errorf("expected %v in trace: %v", gcRanRE, trace)
+		if trace := gcTable(false /* skipShouldQueue */); !nonZeroProgressRE.MatchString(trace) {
+			return fmt.Errorf("expected %v in trace: %v", nonZeroProgressRE, trace)
 		}
 		return nil
 	})

